### ISSUE TITLE
flang fails to build on AArch64 with error: '(((const char*)(& vaArgs)) == 0)' is not a constant expression

### ISSUE THREAD

**Issue by ohno-fj** [llvm/llvm-project#59853]
I built flang in the same way as usual, but I got an error. Is there a problem with my build method?   

The details are as follows:   
- hash value of the source
```
    $ git show | head -1
    commit 41005b7ab2ece511a441d489e9130b82f4fab4cd
    $
```
- compiler version
```
    $ c++ --version | head -1
    c++ (GCC) 8.3.1 20191121 (Red Hat 8.3.1-5)
    $
```
- built machine
```
    AArch64
```
- options for cmake 
```
    cmake \
      -S ../llvm -B ../build \
      -G Ninja ../llvm \
      -DLLVM_PARALLEL_COMPILE_JOBS=9 \
      -DLLVM_PARALLEL_LINK_JOBS=3 \
      -DCMAKE_BUILD_TYPE=Release \
      -DFLANG_ENABLE_WERROR=On \
      -DLLVM_ENABLE_ASSERTIONS=ON \
      -DLLVM_TARGETS_TO_BUILD=host \
      -DCMAKE_INSTALL_PREFIX=/home/users/release \
      -DLLVM_LIT_ARGS=-v \
      -DLLVM_ENABLE_PROJECTS="mlir;flang;clang;openmp" \
      -DLLVM_ENABLE_RUNTIMES="compiler-rt;libcxx;libcxxabi"
`
```
- log of ninja (The following is an excerpt from the log)
```
FAILED: tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o
/usr/bin/c++ -DFLANG_INCLUDE_TESTS=1 -DFLANG_LITTLE_ENDIAN=1 -DGTEST_HAS_RTTI=0 -D_DEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/users/ea01/ea0178/LLVM_20230105/llvm-project/build/tools/flang/lib/Evaluate -I/home/users/ea01/ea0178/LLVM_20230105/llvm-project/flang/\
lib/Evaluate -I/home/users/ea01/ea0178/LLVM_20230105/llvm-project/flang/include -I/home/users/ea01/ea0178/LLVM_20230105/llvm-project/build/tools/flang/include -I/home/users/ea01/ea0178/LLVM_20230105/llvm-project/build/include -I/home/users/ea01/ea0178/LLVM_20230105/llvm-project/llvm/include -isystem /home/users/ea0\
1/ea0178/LLVM_20230105/llvm-project/llvm/../mlir/include -isystem /home/users/ea01/ea0178/LLVM_20230105/llvm-project/build/tools/mlir/include -isystem /home/users/ea01/ea0178/LLVM_20230105/llvm-project/build/tools/clang/include -isystem /home/users/ea01/ea0178/LLVM_20230105/llvm-project/llvm/../clang/include -fPIC \
-fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-co\
mment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Werror -Wno-deprecated-copy -Wno-ctad-maybe-unsupported -fno-strict-aliasing -fno-semantic-interposition -O3 -DNDEBUG  -fno-exceptions -fno-rtti -UNDEBUG -std=c++17 -MD -MT tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvalu\
ate.dir/constant.cpp.o -MF tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o.d -o tools/flang/lib/Evaluate/CMakeFiles/obj.FortranEvaluate.dir/constant.cpp.o -c /home/users/ea01/ea0178/LLVM_20230105/llvm-project/flang/lib/Evaluate/constant.cpp
In file included from /home/users/llvm-project/flang/include/flang/Common/idioms.h:26,
                 from /home/users/llvm-project/flang/include/flang/Common/indirection.h:23,
                 from /home/users/llvm-project/flang/include/flang/Evaluate/formatting.h:22,
                 from /home/users/llvm-project/flang/include/flang/Evaluate/constant.h:12,
                 from /home/users/ea01/ea0178/LLVM_20230105/llvm-project/flang/lib/Evaluate/constant.cpp:9:
/home/users/llvm-project/flang/include/flang/Common/enum-class.h: In static member function 'static std::string_view Fortran::evaluate::ComplexPart::EnumToString(Fortran::evaluate::ComplexPart::Part)':
/home/users/llvm-project/flang/include/flang/Evaluate/variable.h:357:3:   in 'constexpr' expansion of 'Fortran::common::EnumNames<2>(((const char*)(& vaArgs)))'
/home/users/llvm-project/flang/include/flang/Common/enum-class.h:51:16: error: '(((const char*)(& vaArgs)) == 0)' is not a constant expression
     } else if (!start) {
                ^~~~~~
/home/users/llvm-project/flang/include/flang/Common/enum-class.h: In static member function 'static std::string_view Fortran::evaluate::DescriptorInquiry::EnumToString(Fortran::evaluate::DescriptorInquiry::Field)':
/home/users/llvm-project/flang/include/flang/Evaluate/variable.h:412:3:   in 'constexpr' expansion of 'Fortran::common::EnumNames<5>(((const char*)(& vaArgs)))'
/home/users/llvm-project/flang/include/flang/Common/enum-class.h:51:16: error: '(((const char*)(& vaArgs)) == 0)' is not a constant expression
/home/users/llvm-project/flang/include/flang/Common/enum-class.h: In static member function 'static std::string_view Fortran::semantics::MiscDetails::EnumToString(Fortran::semantics::MiscDetails::Kind)':
/home/users/llvm-project/flang/include/flang/Semantics/symbol.h:373:3:   in 'constexpr' expansion of 'Fortran::common::EnumNames<11>(((const char*)(& vaArgs)))'
/home/users/llvm-project/flang/include/flang/Common/enum-class.h:51:16: error: '(((const char*)(& vaArgs)) == 0)' is not a constant expression
/home/users/llvm-project/flang/include/flang/Common/enum-class.h: In static member function 'static std::string_view Fortran::semantics::GenericKind::EnumToString(Fortran::semantics::GenericKind::OtherKind)':
/home/users/llvm-project/flang/include/flang/Semantics/symbol.h:443:3:   in 'constexpr' expansion of 'Fortran::common::EnumNames<4>(((const char*)(& vaArgs)))'
/home/users/llvm-project/flang/include/flang/Common/enum-class.h:51:16: error: '(((const char*)(& vaArgs)) == 0)' is not a constant expression
/home/users/llvm-project/flang/include/flang/Common/enum-class.h: In static member function 'static std::string_view Fortran::semantics::GenericKind::EnumToString(Fortran::semantics::GenericKind::DefinedIo)':
/home/users/llvm-project/flang/include/flang/Semantics/symbol.h:444:3:   in 'constexpr' expansion of 'Fortran::common::EnumNames<4>(((const char*)(& vaArgs)))'
/home/users/llvm-project/flang/include/flang/Common/enum-class.h:51:16: error: '(((const char*)(& vaArgs)) == 0)' is not a constant expression
/home/users/llvm-project/flang/include/flang/Common/enum-class.h: In static member function 'static std::string_view Fortran::semantics::Symbol::EnumToString(Fortran::semantics::Symbol::Flag)':
/home/users/llvm-project/flang/include/flang/Semantics/symbol.h:526:3:   in 'constexpr' expansion of 'Fortran::common::EnumNames<57>(((const char*)(& vaArgs)))'
/home/users/llvm-project/flang/include/flang/Common/enum-class.h:51:16: error: '(((const char*)(& vaArgs)) == 0)' is not a constant expression
At global scope:
cc1plus: error: unrecognized command line option '-Wno-ctad-maybe-unsupported' [-Werror]
cc1plus: error: unrecognized command line option '-Wno-deprecated-copy' [-Werror]
cc1plus: all warnings being treated as errors
```

**Comment by kawashima-fj** [comment]
This problem my be caused by a GCC bug fixed in GCC 8.4.

[I extracted related code](https://github.com/llvm/llvm-project/files/10403190/issue-59853-small.txt) from `flang/include/flang/Common/enum-class.h` and `flang/include/flang/Evaluate/variable.h`. [godbolt](https://godbolt.org/) shows [the same compilation error with g++ 8.3](https://godbolt.org/z/PsPj415ba) and [no error with g++ 8.4](https://godbolt.org/z/YE5vYEbGY).

I don't know which [bug fix](https://gcc.gnu.org/bugzilla/buglist.cgi?bug_status=RESOLVED&component=c%2B%2B&resolution=FIXED&target_milestone=8.4) affects.

**Comment by ohno-fj** [comment]
@kawashima-fj  Thanks for your comment

I built the main branch with g + + 8.4, and I was able to build flang without any problems.
For future builds, I'll use GCC 8.4.
